### ISSUE TITLE
Spark hanging on docker-machine 0.5.1

### ISSUE THREAD

**Issue by dougmet** [jupyter/docker-stacks#75]
I'm having some trouble with both pyspark-notebook and all-spark-notebook on version 0.5.1 of docker-machine. It's working fine for me on an earlier version, 0.4.1 and that's the only system difference.

When I run the test code to see it's all up and running:
```python
import pyspark
sc = pyspark.SparkContext('local[*]')
```
It hangs on trying to load the context. This also eventually brings down the whole machine.

Anyone else seen this on 0.5.1?

**Comment by parente** [comment]
Are you using virtualbox or a remote VM? What version of virtualbox if local? And do you see any logs in the notebook server console from spark?

I'm on docker-machine 0.5.1, virtualbox 4.3.32, and the latest (as of this morning) pyspark-notebook image. Couldn't reproduce it here.

![screen shot 2015-12-08 at 11 29 02 am](https://cloud.githubusercontent.com/assets/153745/11661252/ed3210c0-9d9e-11e5-88f5-7e4c3e9b725a.png)

![screen shot 2015-12-08 at 11 29 06 am](https://cloud.githubusercontent.com/assets/153745/11661257/f0bc645c-9d9e-11e5-8b8b-abef47c06eb5.png)

**Comment by dougmet** [comment]
Good questions. I've got virtual box 5.0.10. I couldn't find any logs but I'll look into that. Thanks for trying to reproduce, if it's just me I'll close.

**Comment by dougmet** [comment]
Launching the notebook from the console has defeated me so I can't see the logs. Thanks for looking at it, feel free to close.

**Comment by parente** [comment]
Out of curiosity, how were you launching it? Are you using Kitematic?

**Comment by dougmet** [comment]
Launching from command line (Mac):
```
docker run -d -p 8888:8888 jupyter/all-spark-notebook
```

Then later I tried to go in just with bash so I could see the logging:
```
docker run -it -p 8888:8888 jupyter/all-spark-notebook /bin/bash
```
ad then ran `jupyter notebook` to start the server (which didn't seem to serve it back, which was strange).

**Comment by parente** [comment]
Try:

```
docker run --rm -it -p 8888:8888 jupyter/all-spark-notebook
```

which will keep the container in the foreground so you can see all the logs. Ctrl-C to kill it when done.

(Or, alternatively:

```
docker run --name notebook -d -p 8888:8888 jupyter/all-spark-notebook
```

and then:

```
docker logs notebook
```
)

**Comment by dougmet** [comment]
Thanks! OK, so this is what I get. The cell never finishes executing:
![image](https://cloud.githubusercontent.com/assets/5878305/11690100/ee764160-9e8c-11e5-99ce-9d4444b4887d.png)

and the console output does not do much:
![image](https://cloud.githubusercontent.com/assets/5878305/11690137/21c0f7d6-9e8d-11e5-987b-bf9a117b1df7.png)

The notebook itself keeps running (you can see the checkpoint save coming in a couple of minutes after the start). Bit of a head scratcher.

**Comment by parente** [comment]
Hmm. Yes. Nothing at all happening there. Silly question, but does `x = 1` work if you run it (without the pyspark bit). Just confirming it's a pyspark issue specifically.

Another thing to try: take out the `'local[*]'` value and instantiate the context without any parameters. Any difference?

**Comment by dougmet** [comment]
A reasonable question! Yes, the Python session is working (so, x=1 and then outputting it).

It's still hanging when I run
```
sc = pyspark.SparkContext()
```

I know some people have reported issues with virtual box 5 and networking. Could it possibly be something in that area?

**Comment by parente** [comment]
Maybe, but it seems unlikely. I'd at least expect spark to start logging *something* and then fail a bit later if it were networking. And even then, the Spark comm is entirely within container so it's a little hard to imagine how networking would be impacting it.

**Comment by nicopresto** [comment]
@dougmet @parente I'm having the same issue with SparkContext hanging and bringing down the VM with pyspark-notebook and all-spark-notebook.

Mac OSX Yosemite 10.10.5
VirtualBox Version 5.0.10 
Docker version 1.9.1

**Comment by parente** [comment]
I took the plunge and am now running:

Mac OSX Yosemite 10.10.5
VirtualBox Version 5.0.10 r104061
Docker version 1.9.1

plus the latest `jupyter/pyspark-notebook`

```
$ docker pull jupyter/pyspark-notebook
Using default tag: latest
latest: Pulling from jupyter/pyspark-notebook
283ca018ca09: Pull complete
7171a7619d25: Pull complete
caebc71f9562: Pull complete
c211d6f05a49: Pull complete
Digest: sha256:80da706197a2429fe722f0abdcbbac86f5ad7e5b5c88887deb8a08094f7fa563
Status: Downloaded newer image for jupyter/pyspark-notebook:latest

$ docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
I 22:08:37.794 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 22:08:37.874 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[W 22:08:37.874 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using authentication. This is highly insecure and not recommended.
[I 22:08:37.881 NotebookApp] Serving notebooks from local directory: /home/jovyan/work
[I 22:08:37.881 NotebookApp] 0 active kernels
[I 22:08:37.881 NotebookApp] The IPython Notebook is running at: http://[all ip addresses on your system]:8888/
[I 22:08:37.882 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
^[[O^[[I^[[O[I 22:08:55.928 NotebookApp] 302 GET / (192.168.99.1) 1.22ms
[I 22:08:58.681 NotebookApp] Creating new notebook in
[I 22:08:58.686 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[I 22:08:59.946 NotebookApp] Kernel started: 55d8f758-a223-42ce-9205-521f300bc7d3
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
15/12/11 22:09:13 INFO SparkContext: Running Spark version 1.5.1
15/12/11 22:09:13 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
15/12/11 22:09:14 INFO SecurityManager: Changing view acls to: jovyan
15/12/11 22:09:14 INFO SecurityManager: Changing modify acls to: jovyan
15/12/11 22:09:14 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: Set(jovyan); users with modify permissions: Set(jovyan)
^[[I^[[O15/12/11 22:09:14 INFO Slf4jLogger: Slf4jLogger started
15/12/11 22:09:14 INFO Remoting: Starting remoting
15/12/11 22:09:15 INFO Remoting: Remoting started; listening on addresses :[akka.tcp://sparkDriver@172.17.0.2:52562]
15/12/11 22:09:15 INFO Utils: Successfully started service 'sparkDriver' on port 52562.
15/12/11 22:09:15 INFO SparkEnv: Registering MapOutputTracker
15/12/11 22:09:15 INFO SparkEnv: Registering BlockManagerMaster
15/12/11 22:09:15 INFO DiskBlockManager: Created local directory at /tmp/blockmgr-4a19ac5d-2ee3-4b75-a8b6-d409f526d70b
15/12/11 22:09:15 INFO MemoryStore: MemoryStore started with capacity 530.3 MB
...
```

<img width="664" alt="screen shot 2015-12-11 at 5 11 43 pm" src="https://cloud.githubusercontent.com/assets/153745/11756758/45f49ad8-a02a-11e5-85f4-d751d2a94383.png">

What VM are you both running in VirtualBox? I used docker-machine 0.5.1 to provision the boot2docker image.

**Comment by nicopresto** [comment]
Name            | Ver
--------------- | --------------------
Docker Toolbox  | 1.9.1c
Docker-machine  | 0.5.2 ( 0456b9f )
Docker          | 1.9.1, build a34a1d5
Mac OSX         | 10.10.5
VirtualBox      | 5.0.10 

```
$ docker pull jupyter/pyspark-notebook
Using default tag: latest
latest: Pulling from jupyter/pyspark-notebook
Digest: sha256:80da706197a2429fe722f0abdcbbac86f5ad7e5b5c88887deb8a08094f7fa563
Status: Image is up to date for jupyter/pyspark-notebook:latest

$ docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
[I 00:09:56.362 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 00:09:56.567 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[W 00:09:56.568 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using authentication. This is highly insecure and not recommended.
[I 00:09:56.597 NotebookApp] Serving notebooks from local directory: /home/jovyan/work
[I 00:09:56.598 NotebookApp] 0 active kernels 
[I 00:09:56.598 NotebookApp] The IPython Notebook is running at: http://[all ip addresses on your system]:8888/
[I 00:09:56.599 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 00:10:27.151 NotebookApp] 302 GET / (192.168.99.1) 1.71ms
[I 00:10:32.600 NotebookApp] Creating new notebook in 
[I 00:10:32.607 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[I 00:10:34.526 NotebookApp] Kernel started: 4c0b3c83-9589-4796-85bf-a26f32849742
[I 00:12:34.532 NotebookApp] Saving file at /test.ipynb
[I 00:15:04.898 NotebookApp] Saving file at /test.ipynb
```

![image](https://cloud.githubusercontent.com/assets/480671/11758635/8a17ef92-a03b-11e5-963d-219bec16c47b.png)

I used the latest docker toolbox. The SparkContext step hangs and I can't stop the container. The only way to resume is to restart the VM via Virtualbox. 

Looks like our docker-machine versions are different (0.5.2 vs 0.5.1). I'll try to revert my docker-machine and retest.

**Comment by nicopresto** [comment]
This is the message when I interrupt the kernel:

```
$ docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
[I 00:09:56.362 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 00:09:56.567 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[W 00:09:56.568 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using authentication. This is highly insecure and not recommended.
[I 00:09:56.597 NotebookApp] Serving notebooks from local directory: /home/jovyan/work
[I 00:09:56.598 NotebookApp] 0 active kernels 
[I 00:09:56.598 NotebookApp] The IPython Notebook is running at: http://[all ip addresses on your system]:8888/
[I 00:09:56.599 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 00:10:27.151 NotebookApp] 302 GET / (192.168.99.1) 1.71ms
[I 00:10:32.600 NotebookApp] Creating new notebook in 
[I 00:10:32.607 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[I 00:10:34.526 NotebookApp] Kernel started: 4c0b3c83-9589-4796-85bf-a26f32849742
[I 00:12:34.532 NotebookApp] Saving file at /test.ipynb
[I 00:15:04.898 NotebookApp] Saving file at /test.ipynb
[I 00:23:06.445 NotebookApp] Kernel interrupted: 4c0b3c83-9589-4796-85bf-a26f32849742
[I 00:24:34.554 NotebookApp] Saving file at /test.ipynb
```

![image](https://cloud.githubusercontent.com/assets/480671/11758751/f8ca84d0-a03c-11e5-9f2c-346b3516157e.png)

**Comment by nicopresto** [comment]
I used the follow command to provision the virtual machine:

    $ docker-machine create lab -d=virtualbox

**Comment by parente** [comment]
While it's hanging, what do you see if you run `ps auxf` in a Jupyter terminal tab? It looks like either the JVM process that pyspark uses under the covers is not launching or the kernel cannot communicate with it. Both seem very odd.

**Comment by nicopresto** [comment]
root@338f1d750f3e:/home/jovyan/work# ps auxf
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0   4212   660 ?        Ss   04:18   0:00 tini -- /bin/bash
root         8  0.0  0.1  20240  3208 ?        S    04:18   0:00 /bin/bash
root        32  0.0  0.0  17488  2108 ?        R+   04:20   0:00  \_ ps auxf

**Comment by dougmet** [comment]
Thanks @parente for looking at this in such detail, and @nicopresto for helping out. I'm running docker-machine version 0.5.1 (7e8e38e) so the only difference between our setups is that I'm in 10.11.1 (El Capitan). Seems hard to believe that the host could make much of a difference though.

**Comment by nicopresto** [comment]
Not sure if helpful, but I tried a couple more things:
1) Reverted to docker-machine version 0.5.1 (7e8e38e) and encountered the same problem.
2) Tried to launch pyspark and spark-shell from Jupyter terminal and encountered the same problem.

Process report before launch:
```
$ ps auxf                                                                                                
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND                                                           
root         1  0.0  0.0   4212   616 pts/0    Ss+  16:16   0:00 tini -- start-notebook.sh                                         
root         7  0.0  0.0  44744  2564 pts/0    S+   16:16   0:00 su jovyan -c env PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bi
jovyan       9  0.4  1.6 337472 43688 ?        Ssl  16:16   0:01  \_ /opt/conda/bin/python /opt/conda/bin/jupyter-notebook         
jovyan      18  0.2  1.2 547924 33940 ?        Sl   16:16   0:00      \_ /opt/conda/bin/python -m ipykernel -f /home/jovyan/.local/
jovyan      30  2.8  0.1  20256  3232 pts/0    Ss   16:23   0:00      \_ /bin/bash                                                 
jovyan      33  0.0  0.0  17488  2008 pts/0    R+   16:24   0:00          \_ ps auxf 
```

Process reports during "hang"
```
$ ps auxf                                                                                                
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND                                                           
root         1  0.0  0.0   4212   616 pts/0    Ss+  16:16   0:00 tini -- start-notebook.sh                                         
root         7  0.0  0.0  44744  2564 pts/0    S+   16:16   0:00 su jovyan -c env PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bi
jovyan       9  0.3  1.6 337472 43912 ?        Ssl  16:16   0:02  \_ /opt/conda/bin/python /opt/conda/bin/jupyter-notebook         
jovyan      18  0.1  1.2 547924 33940 ?        Sl   16:16   0:00      \_ /opt/conda/bin/python -m ipykernel -f /home/jovyan/.local/
jovyan      30  0.0  0.1  20256  3236 pts/0    Ss   16:23   0:00      \_ /bin/bash                                                 
jovyan      78  0.0  0.0  11652  2592 pts/0    S+   16:25   0:00      |   \_ bash ./spark-shell                                    
jovyan      85  0.0  0.0  11644  2560 pts/0    S+   16:25   0:00      |       \_ bash /usr/local/spark/bin/spark-class org.apache.s
jovyan     106  0.0  0.0  11644  1592 pts/0    S+   16:25   0:00      |           \_ bash /usr/local/spark/bin/spark-class org.apac
jovyan     107 49.4  0.0      0     0 pts/0    Zl+  16:25   0:30      |               \_ [java] <defunct>                          
jovyan     117  0.3  0.1  20256  3264 pts/1    Ss   16:25   0:00      \_ /bin/bash                                                 
jovyan     121  0.0  0.0  17488  2032 pts/1    R+   16:26   0:00          \_ ps auxf                                               
jovyan      67  0.0  0.0  11644  1592 pts/0    S    16:25   0:00 bash /usr/local/spark/bin/spark-class org.apache.spark.deploy.Spar
jovyan      68 63.0  0.0      0     0 pts/0    Zl   16:25   0:53  \_ [java] <defunct>     
```

Process reports after interrupting pyspark and spark-shell kernels:
```
jovyan@92f62ec50263:/usr/local/spark/bin$ ./pyspark                                                                                
^C                                                                                                                                 
jovyan@92f62ec50263:/usr/local/spark/bin$ ./spark-shell                                                                            
^C 
$ ps auxf                                                                                                
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND                                                           
root         1  0.0  0.0   4212   616 pts/0    Ss+  16:16   0:00 tini -- start-notebook.sh                                         
root         7  0.0  0.0  44744  2564 pts/0    S+   16:16   0:00 su jovyan -c env PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bi
jovyan       9  0.3  1.6 337472 43912 ?        Ssl  16:16   0:02  \_ /opt/conda/bin/python /opt/conda/bin/jupyter-notebook         
jovyan      18  0.1  1.2 547924 33940 ?        Sl   16:16   0:00      \_ /opt/conda/bin/python -m ipykernel -f /home/jovyan/.local/
jovyan      30  0.0  0.1  20256  3236 pts/0    Ss+  16:23   0:00      \_ /bin/bash                                                 
jovyan     117  0.1  0.1  20256  3264 pts/1    Ss   16:25   0:00      \_ /bin/bash                                                 
jovyan     124  0.0  0.0  17488  2068 pts/1    R+   16:28   0:00          \_ ps auxf                                               
jovyan      67  0.0  0.0  11644  1592 pts/0    S    16:25   0:00 bash /usr/local/spark/bin/spark-class org.apache.spark.deploy.Spar
jovyan      68 55.7  0.0      0     0 pts/0    Zl   16:25   1:49  \_ [java] <defunct>                                              
jovyan     106  0.0  0.0  11644  1592 pts/0    S    16:25   0:00 bash /usr/local/spark/bin/spark-class org.apache.spark.deploy.Spar
jovyan     107 49.9  0.0      0     0 pts/0    Zl   16:25   1:26  \_ [java] <defunct>                                              
```                                                

Is it possible that this issue is related to PID reaping, despite using tini?
http://jupyter-notebook.readthedocs.org/en/latest/public_server.html#docker-cmd

**Comment by nicopresto** [comment]
While testing the shell commands above, that instance worked when I retested the Jupyter notebook I'd left open. I've not been able to recreate the success, but here is are the logs. The only difference I can see is the 302:

    302 GET / (192.168.99.1) 1.10ms

Worked:
```
docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
[I 16:16:07.981 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 16:16:08.178 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[W 16:16:08.179 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using authentication. This is highly insecure and not recommended.
[I 16:16:08.210 NotebookApp] Serving notebooks from local directory: /home/jovyan/work
[I 16:16:08.211 NotebookApp] 0 active kernels
[I 16:16:08.211 NotebookApp] The IPython Notebook is running at: http://[all ip addresses on your system]:8888/
[I 16:16:08.212 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 16:16:15.163 NotebookApp] 302 GET / (192.168.99.1) 1.10ms
[I 16:16:21.036 NotebookApp] Creating new notebook in
[I 16:16:21.043 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[I 16:16:22.861 NotebookApp] Kernel started: 601c03d0-e566-4259-b3e4-98c38acd70df
[I 16:18:22.823 NotebookApp] Saving file at /Untitled.ipynb
[I 16:23:57.428 NotebookApp] New terminal with specified name: 1
[I 16:25:57.357 NotebookApp] New terminal with automatic name: 2
[I 16:58:23.531 NotebookApp] Saving file at /Untitled.ipynb
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
15/12/12 17:01:39 INFO SparkContext: Running Spark version 1.5.1
```

Failed:
```
docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
[I 17:35:07.134 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 17:35:07.317 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[W 17:35:07.318 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using authentication. This is highly insecure and not recommended.
[I 17:35:07.342 NotebookApp] Serving notebooks from local directory: /home/jovyan/work
[I 17:35:07.343 NotebookApp] 0 active kernels
[I 17:35:07.344 NotebookApp] The IPython Notebook is running at: http://[all ip addresses on your system]:8888/
[I 17:35:07.345 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 17:35:25.349 NotebookApp] Creating new notebook in
[I 17:35:25.356 NotebookApp] Writing notebook-signing key to /home/jovyan/.local/share/jupyter/notebook_secret
[I 17:35:27.814 NotebookApp] Kernel started: dc731b08-e186-4bc6-a72b-72418a873637
[I 17:35:39.048 NotebookApp] Saving file at /Untitled.ipynb
[I 17:35:58.672 NotebookApp] New terminal with automatic name: 1
[I 17:36:04.322 NotebookApp] New terminal with automatic name: 2
[I 17:49:27.853 NotebookApp] Saving file at /Untitled.ipynb
[I 17:50:40.816 NotebookApp] Kernel interrupted: dc731b08-e186-4bc6-a72b-72418a873637
```

**Comment by parente** [comment]
Sorry for all the questions back but here's one more based on the ps: how much RAM do you have allocated to the VM? I wonder if the local Spark driver cannot get enough and so the JVM immediately dies.

http://spark.apache.org/docs/latest/configuration.html#available-properties

**Comment by dougmet** [comment]
2 GB for me. That's the same on this host and the host where it's working. Do you think that's enough?

**Comment by nicopresto** [comment]
Thanks for the suggestion to increase memory, alas no such luck. 

I started with a fresh 4GB vm to retest, here are the steps:

```
docker-machine create notebooks -d=virtualbox --virtualbox-memory "4096"
docker-machine env notebooks
eval "$(docker-machine env notebooks)"
docker pull jupyter/pyspark-notebook
docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
```

Test in browser (Jupyter notebook)

```
import pyspark
sc = pyspark.SparkContext()
```

... hangs

![image](https://cloud.githubusercontent.com/assets/480671/11768393/c1e73694-a199-11e5-8290-084ac201d8fa.png)

**Comment by parente** [comment]
Whelp, I'm stumped and fresh out of ideas.

@jtyberg @lbustelo either of you having issues with pyspark in the latest pyspark-notebook or all-spark-notebook?

**Comment by parente** [comment]
@nicopresto @dougmet  grasping at straws now but would either of you happen to be running a VPN client on your host box?

**Comment by jtyberg** [comment]
I was able to reproduce the problem on a local VM on my Mac.

Mac OSX Yosemite 10.10.5
VirtualBox Version 5.0.10 r104061

```
$ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      linux/amd64
$ docker-machine version
docker-machine version 0.5.2 ( 0456b9f )
```

```
$ docker pull jupyter/pyspark-notebook
$ docker images
REPOSITORY                 TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
jupyter/pyspark-notebook   latest              c211d6f05a49        10 days ago         3.919 GB
```

```
$ docker-machine ssh dev
...
Boot2Docker version 1.9.1, build master : cef800b - Fri Nov 20 19:33:59 UTC 2015
Docker version 1.9.1, build a34a1d5
```

@parente and I tracked it down to a problem running the `/usr/local/spark/bin/spark-class` script within the container.  Essentially, invoking this command works the first time, but fails every time thereafter, even if running it from a *different* container, so it's definitely a container/VM thing.

```
java -cp /usr/local/spark/lib/spark-assembly-1.5.1-hadoop2.6.0.jar org.apache.spark.launcher.Main org.apache.spark.deploy.SparkSubmit --class org.apache.spark.repl.Main --name Spark shell
```

When this cmd hangs, the `spark-shell` script never completes.

**Comment by parente** [comment]
We did a bit more sleuthing in the docker issues and found docker/docker#18180 which shows lots of people having java defunct processes on the newer boot2docker VM builds used by virtual machine.

So, a few work arounds:

- I don't hit this problem on `Boot2Docker version 1.9.0, build master : 16e4a2a`. Downgrading the boot2docker.iso used by docker-machine to this version might work. (I don't know of an official way to do that other than finding it on disk and overwriting it.)
- The issue above suggests adding multiple virtual CPUs to the virtualbox image helps, but does not resolve the problem entirely.

**Comment by parente** [comment]
All boot2docker.iso releases are here: https://github.com/boot2docker/boot2docker/releases

**Comment by parente** [comment]
And it looks like you can swap in the older boot2docker.iso in `~/.docker/machine/cache` to get docker-machine to use it on next VM creation.

**Comment by jtyberg** [comment]
Reverting to boot2docker.iso 1.9.0 fixed the problem for me.

**Comment by parente** [comment]
:beers: for @jtyberg for sitting with me on plowing through this (I'd do :scotches: but, darn, no emoji)

**Comment by nicopresto** [comment]
Great! Thanks all for isolating the problem as a VM issue and proposing a workaround. 

I followed your suggestion to roll back the boot2docker version and it worked. Thanks especially for identifying the issue of the spark-class running the first time, and failing thereafter. I was perplexed trying to reproduce one successful run and wonder if related.

Here were the steps:

Replaced boot2docker.iso on host machine

    ~/.docker/machine/cache
![image](https://cloud.githubusercontent.com/assets/480671/11828870/3dad9f34-a366-11e5-9db6-f0d42401c9cf.png)
with boot2docker v1.9.0 from:

    https://github.com/boot2docker/boot2docker/releases/tag/v1.9.0

Built new vm using boot2docker v1.9.0

```
    docker-machine create notebooks -d=virtualbox
    docker-machine env notebooks
    eval "$(docker-machine env notebooks)"
    docker pull jupyter/pyspark-notebook
    docker run -it --rm -p 8888:8888 jupyter/pyspark-notebook
```

Test in browser (Jupyter notebook)

```
    import pyspark
    sc = pyspark.SparkContext()
    sc.parallelize([1,2,3]).take(1)
```

![image](https://cloud.githubusercontent.com/assets/480671/11828795/cd826474-a365-11e5-8d42-7eea022a5888.png)

Cheers

**Comment by dougmet** [comment]
All: thanks so much for all the effort finding the problem. Really appreciate it.

**Comment by finleysg** [comment]
At least from Windows, I found that I had to specify the 1.9.0 version with the --virtualbox-boot2docker-url switch:

```
docker-machine create -d virtualbox --virtualbox-boot2docker-url=https://github.com/boot2docker/boot2docker/releases/download/v1.9.0/boot2docker.iso pyspark
```

**Comment by schzcas** [comment]
Hi all!

I encountered this problem a few days ago and I was doing a lot of bunch of stuff.
The problem was the same...It hangs when I run sc = pyspark.SparkContext()

I have OS X El Capitan 10.11.3
When I went to see the VirtualBox application (Oracle) that is needed in mac to run docker, it asked if a wanted to update to version 5.0.16
Also, I found that I was running an old version of Docker, so I also thought to update it.

So pretty much, I took the kamikaze strategy. ERASE EVERYTHING AND INSTALL IT AGAIN. 
It about 30 mins. 

voilá, now everything works perfectly
